pydantic_ai_slim/pydantic_ai/agent.py

```diff
@@ -737,62 +752,158 @@ async def _handle_model_response(
+        # Check if we allow plain text results
+        if not self._allow_text_result:
+            self._incr_result_retry()
+            response = _messages.RetryPrompt(
+                content='Plain text responses are not permitted, please call one of the functions instead.',
+            )
+            return None, [response]
+
+        # Try to validate the text response as a result
+        result_data_input = cast(ResultData, model_response.content)
+        try:
+            result_data = await self._validate_result(result_data_input, deps, None)
+        except _result.ToolRetryError as e:
+            self._incr_result_retry()
+            return None, [e.tool_retry]
+
+        return _MarkFinalResult(result_data), []
+
+    async def _handle_structured_response(
+        self, model_response: _messages.ModelStructuredResponse, deps: AgentDeps
+    ) -> tuple[_MarkFinalResult[ResultData] | None, list[_messages.Message]]:
+        """Handle a structured response containing tool calls from the model."""
+        if not model_response.calls:
+            raise exceptions.UnexpectedModelBehavior('Received empty tool call message')
+
+        # First process any final result tool calls
+        final_result, final_messages = await self._process_final_tool_calls(model_response, deps)
+
+        # Then process regular tools based on end strategy
+        if not (final_result and self.end_strategy == 'early'):
+            tool_messages = await self._process_regular_tool_calls(model_response, deps)
+        else:
+            tool_messages = self._mark_skipped_regular_tools(model_response)
+
+        return final_result, [*final_messages, *tool_messages]
+
+    async def _process_final_tool_calls(
+        self,
+        model_response: _messages.ModelStructuredResponse,
+        deps: AgentDeps,
+    ) -> tuple[_MarkFinalResult[ResultData] | None, list[_messages.Message]]:
+        """Process any final result tool calls and return the first valid result."""
+        if not self._result_schema:
+            return None, []
+
+        messages: list[_messages.Message] = []
+        final_result = None
+
+        for call in model_response.calls:
+            # Skip non-final result tools
+            if call.tool_name not in self._result_schema.tool_names():
+                continue
+
+            # Try to match the call to a result tool
+            match = self._result_schema.find_tool(model_response)
+            if not match:
+                continue
+
+            call_, result_tool = match
+            # Ensure we're processing the right call
+            if call_.tool_name != call.tool_name:
+                continue
+
+            if final_result is None:
+                # This is the first result tool - try to use it
+                try:
+                    result_data = result_tool.validate(call)
+                    result_data = await self._validate_result(result_data, deps, call)
```

Comment:
I suspect we need to refactor this to run these in parallel. I can do that in a separate PR if you like.

Reply:
ignore this, we only run `_validate_result` once anyway.